### PR TITLE
MEN-3996 ws/shell for Remote Terminal.

### PR DESCRIPTION
**Pull request by merlin-northern**
ChangeLog:none
Signed-off-by: Peter Grzybowski <peter@northern.tech>

### CODE REVIEW

File: ws/protomsg.go [hunk 1]
```diff
@@ -28,14 +28,32 @@ const (
...
+	// * MenderConfiguration for configuration of devices
+	// * MenderMonitoring for the stats and monitoring
 	Proto ProtoType `msgpack:"proto"`
+
 	// MsgType is an optional content type header describing
-	// the protocol specific content type of the message.
+	// the protocol specific content type of the message. E.g.: for MenderShell:
+	// * CommandData for the stdout / stderr coming from the terminal and keystrokes from the UI
+	// * StartSession for the session start
+	// * StopSession for stopping of the session
```
Comment (alfrunes) [inline ws/protomsg.go:41]: This is `shell` specific. This package is supposed to be generic.

File: ws/protomsg.go [hunk 2]
```diff
@@ -28,14 +28,32 @@ const (
 // to which protocol the message should be routed.
 type ProtoHdr struct {
 	// Proto defines which protocol this message belongs
-	// to (required).
+	// to (required). E.g.:
+	// * MenderShell for remote terminal
+	// * MenderConfiguration for configuration of devices
+	// * MenderMonitoring for the stats and monitoring
```
Comment (alfrunes) [inline ws/protomsg.go:34]: This also seems superfluous as the documentation for the different types should be placed by their definition. (Notice that this field is of `ProtoType`).

File: ws/protomsg.go [hunk 3]
```diff
@@ -28,14 +28,32 @@ const (
...
 	MsgType string `msgpack:"typ,omitempty"`
+
+	// Status is the indicator of whether the response contains error or other
+	// conditions (required).
+	// E.g.: for the remote terminal case a StartSession MsgType maye result in
+	//       an error due to mas session limit reached.
+	// * Normal message carries a data from successful processing
+	// * Error the given message returned error
+	// * ProtocolError higher level protocol error
+	Status int `msgpack:"state"`
```
Comment (alfrunes) [inline ws/protomsg.go:51]: I'd argue this should be in the `Properties`. The reason being that a very small set of `MsgType` for a given `Proto` would actually set this field, and for some future `Proto`s a status might not make sense at all.

File: ws/register.go [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+	ErrUnsupportedTypeError = errors.New("unsupported type")
+	ErrUnregisteredProtocol = errors.New("protocol not registered")
+)
+
+type Protocol interface {
+	DeEncapsulate(m *ProtoMsg) (interface{}, error)
+	Encapsulate(m interface{}) (*ProtoMsg, error)
+}
+
+var protoMap = map[ProtoType]Protocol{}
```
Comment (alfrunes) [inline ws/register.go:31]: I don't understand the purpose of this interface. All messages holding `shell` data should be of `ProtoTypeShell`, thus this map holds at most one value.

File: ws/shell/shell.go [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+const version = "1.0"
+
+type MenderShellMessageStatus int
+
+const (
+	NormalMessage MenderShellMessageStatus = iota
+	ErrorMessage
+)
+
+var (
```
Comment (alfrunes) [inline ws/shell/shell.go:30]:
```suggestion
const (
```

File: ws/shell/shell.go [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+	SessionId string `json:"session_id" msgpack:"session_id"`
+	//message status, currently normal and error message types are supported
+	Status MenderShellMessageStatus `json:"status_code" msgpack:"status_code"`
+	//the message payload, if
+	// * .Type===MessageTypeShellCommand interpreted as keystrokes and passed
+	//   to the stdin of the terminal running the shell.
+	// * .Type===MessageTypeSpawnShell interpreted as user_id and passed
+	//   to the session.NewMenderShellSession.
+	Data []byte `json:"data" msgpack:"data"`
+}
```
Comment (alfrunes) [inline ws/shell/shell.go:51]: This is overlapping with the `ProtoMsg` definition.
* `Type` <-> `Header.MsgType`
* `SessionId` <-> `Header.SessionID`
* `Data` <-> `Body`

File: ws/shell/shell.go [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+
+type MenderShellProtocol struct {
+	version string
+}
```
Comment (alfrunes) [inline ws/shell/shell.go:55]: Again, I don't understand the purpose of this interface. Why can't we use `ProtoMsg`?
We discussed in the design document that a `Version` could be given as a `ProtoMsg.Header.Properties`.

**Review by alfrunes — COMMENTED** [review]
To be honest, this wasn't quite what I had in mind.
My intention with the `ProtoMsg` definition is that we have a common and minimalistic application layer (on top of websocket) for all future kinds of messages that we communicate over a websocket.
So, for all `shell` specific messages (`ProtoMsg.Header.Proto == ProtoTypeShell` should apply some protocol specific logic for decoding the message body and what fields and properties should be set in the header.
Thus, what I was expecting in the `shell` sub-package are declarations of different `MsgType` headers, and a struct definition for each `MsgType` (that requires a message body) for encoding/decoding the `ProtoMsg.Body`.

**Comment by merlin-northern** [comment]
> To be honest, this wasn't quite what I had in mind.
> My intention with the `ProtoMsg` definition is that we have a common and minimalistic application layer (on top of websocket) for all future kinds of messages that we communicate over a websocket.
> So, for all `shell` specific messages (`ProtoMsg.Header.Proto == ProtoTypeShell` should apply some protocol specific logic for decoding the message body and what fields and properties should be set in the header.
> Thus, what I was expecting in the `shell` sub-package are declarations of different `MsgType` headers, and a struct definition for each `MsgType` (that requires a message body) for encoding/decoding the `ProtoMsg.Body`.

basically same here. let me think on how to restructure to make it more to your taste :)

**Comment by tranchitella** [comment]
FWIW, I fully agree with what @alfrunes said. Let's try to keep this library high-level, generic, without any specific use-case implementation (shell, in this case).

**Pull request closed without merging**